File: src/main/agents/scanner.ts
import { promises as fs, constants as fsc } from 'fs'
import { basename, extname, join } from 'path'
import type { AgentMeta } from '../../shared/scheduler'

const SUPPORTED_EXTS = new Set(['.sh', '.bash', '.zsh', ''])
const DEFAULT_SECTION = 'Agents'

export const META_SUFFIX = '.meta.json'

export type AgentEntry = {
  id: string
  scriptPath: string
  section: string
  metaPath: string
  meta: AgentMeta
}

/**
 * Walks <userData>/agents/. Top-level scripts get section "Agents"; scripts
 * inside a first-level subdirectory adopt that directory's name as their
 * section. Deeper nesting is ignored. IDs (filename minus extension) must
 * be unique across the whole tree — duplicates are dropped with a console
 * warning so the dashboard never shows two agents with the same id.
 *
 * For each script, a sibling `<id>.meta.json` is read if present and folded
 * in as `meta`. Missing or malformed sidecars degrade to an empty meta.
 */
export async function scanScripts(agentsDir: string): Promise<AgentEntry[]> {
  await fs.mkdir(agentsDir, { recursive: true })
  const out: AgentEntry[] = []
  const seen = new Map<string, string>() // id -> first scriptPath we saw

  await collectInto(agentsDir, DEFAULT_SECTION, out, seen)

  let entries: import('fs').Dirent[]
  try {
    entries = await fs.readdir(agentsDir, { withFileTypes: true })
  } catch {
    entries = []
  }
  for (const e of entries) {
    if (!e.isDirectory()) continue
    if (e.name.startsWith('.')) continue
    await collectInto(join(agentsDir, e.name), e.name, out, seen)
  }

  out.sort((a, b) => a.id.localeCompare(b.id))
  return out
}

async function collectInto(
  dir: string,
  section: string,
  out: AgentEntry[],
  seen: Map<string, string>
): Promise<void> {
  let entries: string[]
  try {
    entries = await fs.readdir(dir)
  } catch {
    return
  }
  for (const name of entries) {
    if (name.startsWith('.')) continue
    if (name.toLowerCase() === 'readme.md') continue
    if (name.endsWith(META_SUFFIX)) continue

    const ext = extname(name)
    if (!SUPPORTED_EXTS.has(ext)) continue

    const fullPath = join(dir, name)
    let stat: import('fs').Stats
    try {
      stat = await fs.stat(fullPath)
    } catch {
      continue
    }
    if (!stat.isFile()) continue
    if (!(await isExecutable(fullPath))) continue

    const id = basename(name, ext)
    if (id.startsWith('__')) {
      console.warn(
        `[scanner] reserved agent id "${id}" — the __-prefix is for internal markers; ignoring ${fullPath}`
      )
      continue
    }
    const previous = seen.get(id)
    if (previous) {
      console.warn(
        `[scanner] duplicate agent id "${id}" — keeping ${previous}, ignoring ${fullPath}`
      )
      continue
    }
    seen.set(id, fullPath)
    const metaPath = join(dir, `${id}${META_SUFFIX}`)
    const meta = await readMeta(metaPath)
    out.push({ id, scriptPath: fullPath, section, metaPath, meta })
  }
}

async function readMeta(metaPath: string): Promise<AgentMeta> {
  let txt: string
  try {
    txt = await fs.readFile(metaPath, 'utf8')
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === 'ENOENT') return {}
    console.warn(`[scanner] failed to read ${metaPath}:`, err)
    return {}
  }
  try {
    const parsed = JSON.parse(txt)
    if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) {
      return parsed as AgentMeta
    }
    console.warn(`[scanner] ${metaPath} is not a JSON object — ignoring`)
    return {}
  } catch (err) {
    console.warn(`[scanner] failed to parse ${metaPath}:`, err)
    return {}
  }
}

async function isExecutable(path: string): Promise<boolean> {
  try {
    await fs.access(path, fsc.X_OK)
    return true
  } catch {
    return false
  }
}

export async function ensureAgentsDir(agentsDir: string, seedFrom: string): Promise<void> {
  await fs.mkdir(agentsDir, { recursive: true })
  const existing = await fs.readdir(agentsDir).catch(() => [] as string[])
  if (existing.length > 0) return
  let seedEntries: string[]
  try {
    seedEntries = await fs.readdir(seedFrom)
  } catch {
    return
  }
  for (const name of seedEntries) {
    const src = join(seedFrom, name)
    const dst = join(agentsDir, name)
    try {
      const data = await fs.readFile(src)
      await fs.writeFile(dst, data)
      const stat = await fs.stat(src)
      await fs.chmod(dst, stat.mode)
    } catch {
      /* skip un-readable seeds */
    }
  }
}

File: internal/scheduler/scanner.go
package scheduler

import (
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"golang.org/x/sys/unix"
)

const (
	defaultSection = "Agents"
	metaSuffix     = ".meta.json"
)

var supportedExts = map[string]struct{}{
	".sh":   {},
	".bash": {},
	".zsh":  {},
	"":      {},
}

type Agent struct {
	ID         string
	ScriptPath string
	Section    string
	MetaPath   string
	Meta       AgentMeta
}

type ScanIssue struct {
	Kind string // "duplicate", "not-executable"
	Path string
	Note string
}

type ScanResult struct {
	Agents []Agent
	Issues []ScanIssue
}

// ScanAgents walks agentsDir per the rules in src/main/agents/scanner.ts.
func ScanAgents(agentsDir string) (ScanResult, error) {
	res := ScanResult{}
	if err := os.MkdirAll(agentsDir, 0o755); err != nil {
		return res, err
	}
	seen := map[string]string{}
	if err := collectInto(agentsDir, defaultSection, &res, seen); err != nil {
		return res, err
	}
	dirEntries, err := os.ReadDir(agentsDir)
	if err != nil {
		if !errors.Is(err, os.ErrNotExist) {
			return res, err
		}
		dirEntries = nil
	}
	for _, e := range dirEntries {
		if !e.IsDir() {
			continue
		}
		if strings.HasPrefix(e.Name(), ".") {
			continue
		}
		if err := collectInto(filepath.Join(agentsDir, e.Name()), e.Name(), &res, seen); err != nil {
			return res, err
		}
	}
	sort.Slice(res.Agents, func(i, j int) bool {
		return res.Agents[i].ID < res.Agents[j].ID
	})
	sort.Slice(res.Issues, func(i, j int) bool {
		return res.Issues[i].Path < res.Issues[j].Path
	})
	return res, nil
}

func collectInto(dir, section string, res *ScanResult, seen map[string]string) error {
	entries, err := os.ReadDir(dir)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil
		}
		return err
	}
	for _, e := range entries {
		name := e.Name()
		if strings.HasPrefix(name, ".") {
			continue
		}
		if strings.EqualFold(name, "readme.md") {
			continue
		}
		if strings.HasSuffix(name, metaSuffix) {
			continue
		}
		ext := filepath.Ext(name)
		if _, ok := supportedExts[ext]; !ok {
			continue
		}
		full := filepath.Join(dir, name)
		info, err := os.Stat(full)
		if err != nil || !info.Mode().IsRegular() {
			continue
		}
		if ext == "" {
			ok, _ := hasShebang(full)
			if !ok {
				continue
			}
		}
		if !isExecutable(full) {
			continue
		}
		id := strings.TrimSuffix(name, ext)
		if strings.HasPrefix(id, "__") {
			continue
		}
		if prev, dup := seen[id]; dup {
			res.Issues = append(res.Issues, ScanIssue{
				Kind: "duplicate",
				Path: full,
				Note: fmt.Sprintf("duplicate id %q; kept %s", id, prev),
			})
			continue
		}
		seen[id] = full
		metaPath := filepath.Join(dir, id+metaSuffix)
		meta := readMeta(metaPath)
		res.Agents = append(res.Agents, Agent{
			ID:         id,
			ScriptPath: full,
			Section:    section,
			MetaPath:   metaPath,
			Meta:       meta,
		})
	}
	return nil
}

func isExecutable(path string) bool {
	return unix.Access(path, unix.X_OK) == nil
}

func hasShebang(path string) (bool, error) {
	f, err := os.Open(path)
	if err != nil {
		return false, err
	}
	defer f.Close()
	buf := make([]byte, 2)
	n, err := io.ReadFull(f, buf)
	if err != nil && err != io.ErrUnexpectedEOF {
		if errors.Is(err, io.EOF) {
			return false, nil
		}
		return false, err
	}
	return n == 2 && buf[0] == '#' && buf[1] == '!', nil
}

func readMeta(path string) AgentMeta {
	data, err := os.ReadFile(path)
	if err != nil {
		return AgentMeta{}
	}
	return ParseMeta(data)
}
